### PR TITLE
[Extension]: Grease Quic Bit Support

### LINKED ISSUE #1183: Support Grease QUIC bit Extension
### Describe the feature you'd like supported

This extension (https://tools.ietf.org/html/draft-thomson-quic-bit-grease) allows for the negotiation of greasing the "QUIC bit" or fixed bit in the long and short QUIC packet headers. It would probably be a good idea to support this extension.

### PR DESCRIPTION
**Pull request by liveans**
## Description

Closes #1183. This PR aims to support [draft-ietf-quic-bit-grease-04](https://www.ietf.org/archive/id/draft-ietf-quic-bit-grease-04.html) extension.

## Testing

Necessary tests added to handshake.

## Documentation

Yes, `GreaseQuicBitEnabled` was added to docs.

### CODE REVIEW

File: src/core/quicdef.h [hunk 1]
```diff
@@ -495,6 +495,11 @@ CXPLAT_STATIC_ASSERT(
 //
 #define QUIC_DEFAULT_DEST_CID_UPDATE_IDLE_TIMEOUT_MS 20000
 
+//
+// The default value for enabling grease quic bit extension.
+//
+#define QUIC_DEFAULT_GREASE_QUIC_BIT_ENABLED         TRUE
```
Comment (nibanks) [inline src/core/quicdef.h:501]: I want to (at least for now) disable this by default.

Comment (liveans) [reply src/core/quicdef.h:501]: Alright!

File: src/core/inline.c [hunk 2]
```diff
@@ -517,6 +518,7 @@ QuicPacketEncodeShortHeaderV1(
     _In_ uint16_t BufferLength,
     _Out_writes_bytes_opt_(BufferLength)
         uint8_t* Buffer
+    _In_ BOOLEAN FixedBit
```
Comment (nibanks) [inline src/core/inline.c:521]: But this after the other `_In_ BOOLEAN`

File: src/core/inline.c [hunk 3]
```diff
@@ -502,7 +502,8 @@ QuicPacketEncodeLongHeaderV1(
     _Out_writes_bytes_(BufferLength)
         uint8_t* Buffer,
     _Out_ uint16_t* PayloadLengthOffset,
-    _Out_ uint8_t* PacketNumberLength
+    _Out_ uint8_t* PacketNumberLength,
+    _In_ BOOLEAN FixedBit
```
Comment (nibanks) [inline src/core/inline.c:506]: Put this after the `PacketType`

File: src/core/packet.c [hunk 4]
```diff
@@ -211,7 +211,8 @@ QuicPacketValidateLongHeaderV1(
     _Inout_ CXPLAT_RECV_PACKET* Packet,
     _Outptr_result_buffer_maybenull_(*TokenLength)
         const uint8_t** Token,
-    _Out_ uint16_t* TokenLength
+    _Out_ uint16_t* TokenLength,
+    _In_ BOOLEAN ExpectedFixedBit
```
Comment (nibanks) [inline src/core/packet.c:215]: Not used?

File: src/core/packet.c [hunk 5]
```diff
@@ -565,7 +566,8 @@ _Success_(return != FALSE)
 BOOLEAN
 QuicPacketValidateShortHeaderV1(
     _In_ const void* Owner, // Binding or Connection depending on state
-    _Inout_ CXPLAT_RECV_PACKET* Packet
+    _Inout_ CXPLAT_RECV_PACKET* Packet,
+    _In_ BOOLEAN ExpectedFixedBit
```
Comment (nibanks) [inline src/core/packet.c:570]: Not used?

File: src/core/packet.c (hunk 4 above)
Comment (liveans) [reply src/core/packet.c:215]: I'll use it, just designing it for now.

File: src/core/packet_builder.c [hunk 6]
```diff
@@ -197,6 +197,7 @@ QuicPacketBuilderPrepare(
         Connection->Stats.QuicVersion == QUIC_VERSION_2 ?
             QuicKeyTypeToPacketTypeV2(NewPacketKeyType) :
             QuicKeyTypeToPacketTypeV1(NewPacketKeyType);
+    BOOLEAN GreaseQuicBitTPReceived = (BOOLEAN)((Connection->PeerTransportParams.Flags & QUIC_TP_FLAG_GREASE_QUIC_BIT) > 0);
```
Comment (nibanks) [inline src/core/packet_builder.c:200]:
```suggestion
    BOOLEAN NoGreaseQuicBit = !(Connection->PeerTransportParams.Flags & QUIC_TP_FLAG_GREASE_QUIC_BIT);
```
And then use `NoGreaseQuicBit` below?

File: src/core/binding.c [hunk 7]
```diff
@@ -955,7 +955,7 @@ QuicBindingProcessStatelessOperation(
             PacketLength - QUIC_STATELESS_RESET_TOKEN_LENGTH,
             SendDatagram->Buffer);
         ResetPacket->IsLongHeader = FALSE;
-        ResetPacket->FixedBit = 1;
+        ResetPacket->FixedBit = 1; // TODO : Find a way to check connection settings and PeerTPs.
```
Comment (nibanks) [inline src/core/binding.c:958]: There is no connection state in this case. That's the point of stateless reset. Just always assume grease was disabled.

File: src/core/connection.c [hunk 8]
```diff
@@ -3698,6 +3702,10 @@ QuicConnRecvHeader(
 
     CXPLAT_FRE_ASSERT(QuicIsVersionSupported(Connection->Stats.QuicVersion));
 
+    const BOOLEAN IgnoreFixedBit = (Connection->Settings.GreaseQuicBitEnabled &&
+        (QuicConnIsClient(Connection) ||
```
Comment (nibanks) [inline src/core/connection.c:3706]: Why does client always do it? Put a comment here explaining this?

Comment (nibanks) [reply src/core/connection.c:3706]: And should we constantly be reevaluating this or should we store it on the connection?

Comment (liveans) [reply src/core/connection.c:3706]: I'd prefer to store this state in the connection.

Comment (liveans) [reply src/core/connection.c:3706]: If this feature is turned on in the client, it can accept FixedBit as 0 according to RFC since it will send this TP, but the Server cannot ignore this bit without receiving TP.

File: src/core/binding.c (hunk 7 above)
Comment (liveans) [reply src/core/binding.c:958]: Alright!

File: src/core/connection.c (hunk 8 above)
Comment (liveans) [reply src/core/connection.c:3706]: Yeah, we should definitely store it on the connection. In this way, things will get easier to do.

File: src/core/connection.h [hunk 9]
```diff
@@ -624,6 +624,10 @@ typedef struct QUIC_CONNECTION {
         QUIC_FLOW_BLOCKED_TIMING_TRACKER FlowControl;
     } BlockedTimings;
 
+    uint8_t IgnoreFixedBit : 1;
+    uint8_t FixedBit : 1;
+    uint8_t RESERVED : 6;
```
Comment (nibanks) [inline src/core/connection.h:629]: Put these in the flags/state structure.

File: src/core/crypto_tls.c [hunk 10]
```diff
@@ -1132,6 +1139,24 @@ QuicCryptoTlsEncodeTransportParameters(
...
+        QuicTraceLogConnVerbose(
+            EncodeTPGreaseQuicBit,
+            Connection,
+            "TP: Grease Quic Bit");
+
+        //
+        // An endpoint that advertises the grease_quic_bit transport parameter MUST accept packets
+        // with the QUIC Bit set to a value of 0.
+        //
+        Connection->IgnoreFixedBit = 1;
```
Comment (nibanks) [inline src/core/crypto_tls.c:1158]: Don't actually touch the `Connection` in here. It's used purely for logging.

File: src/core/crypto_tls.c [hunk 11]
```diff
@@ -1780,6 +1805,28 @@ QuicCryptoTlsDecodeTransportParameters(
...
+                    Length,
+                    "Invalid length of QUIC_TP_ID_GREASE_QUIC_BIT");
+                goto Exit;
+            }
+            TransportParams->Flags |= QUIC_TP_FLAG_GREASE_QUIC_BIT;
+            //
+            // Endpoints that receive the grease_quic_bit transport parameter from a peer SHOULD set the QUIC Bit to an unpredictable value
+            // unless another extension assigns specific meaning to the value of the bit.
+            //
+            Connection->FixedBit = 0;
```
Comment (nibanks) [inline src/core/crypto_tls.c:1823]: Don't put this in here.

File: src/core/listener.c [hunk 12]
```diff
@@ -748,6 +748,9 @@ QuicListenerAcceptConnection(
         return;
     }
 
+    Connection->FixedBit = 1;
+    Connection->IgnoreFixedBit = 0;
```
Comment (nibanks) [inline src/core/listener.c:752]: Any non-zero defaults should be set in the connection initialization function. Better still, rework the variables so `0` is the default, so you don't have to do anything special.

Comment (liveans) [reply src/core/listener.c:752]: I assume the Connection Initialization function is `QuicConnAlloc`?

File: src/core/settings.c [hunk 13]
```diff
@@ -1453,6 +1478,7 @@ QuicSettingsSettingsToInternal(
     SETTING_COPY_TO_INTERNAL(MigrationEnabled, Settings, InternalSettings);
     SETTING_COPY_TO_INTERNAL(DatagramReceiveEnabled, Settings, InternalSettings);
     SETTING_COPY_TO_INTERNAL(ServerResumptionLevel, Settings, InternalSettings);
+    SETTING_COPY_TO_INTERNAL(GreaseQuicBitEnabled, Settings, InternalSettings); // We can't copy it via sized version due to bit field operation not allowed on it.
```
Comment (nibanks) [inline src/core/settings.c:1481]: How does this effect down-level support?

File: src/core/listener.c (hunk 12 above)
Comment (nibanks) [reply src/core/listener.c:752]: Yes.

File: src/inc/msquic.h [hunk 14]
```diff
@@ -635,7 +636,8 @@ typedef struct QUIC_SETTINGS {
     uint8_t MigrationEnabled                : 1;
     uint8_t DatagramReceiveEnabled          : 1;
     uint8_t ServerResumptionLevel           : 2;    // QUIC_SERVER_RESUMPTION_LEVEL
-    uint8_t RESERVED                        : 2;
+    uint8_t GreaseQuicBitEnabled            : 1;
```
Comment (nibanks) [inline src/inc/msquic.h:639]: We should probably expose whether this was negotiated or not in the QUIC statistics. Also, we should update the QUIC interop tool to enable this.

Comment (liveans) [reply src/inc/msquic.h:639]: Alright! I'm gonna add some stat fields about it. Also gonna take a look at QUIC Interop tools as well!

File: src/core/settings.c (hunk 13 above)
Comment (liveans) [reply src/core/settings.c:1481]: It shouldn't affect down-level support I suppose, since we're not breaking anything here.

File: src/core/connection.c [hunk 15]
```diff
@@ -1965,6 +1966,8 @@ QuicConnStart(
     Connection->RemoteServerName = ServerName;
     ServerName = NULL;
 
+    Connection->State.IgnoreFixedBit = Configuration->Settings.GreaseQuicBitEnabled;
```
Comment (nibanks) [inline src/core/connection.c:1969]: Do we actually need the `IgnoreFixedBit` flag at all, or can we replace al usage with `Settings.GreaseQuicBitEnabled`?

File: src/core/binding.c [hunk 16]
```diff
@@ -1034,6 +1034,7 @@ QuicBindingProcessStatelessOperation(
         SendDatagram->Length =
             QuicPacketEncodeRetryV1(
                 RecvPacket->LH->Version,
+                TRUE, // TODO : Find a way to check connection settings and PeerTPs.
```
Comment (nibanks) [inline src/core/binding.c:1037]: I think when writing all stateless packets (Retry, Reset) we're going to always set the fixed bit (not grease it) because we're never going to have any connection-specific state to base this decision on. So, really, I think we might as well, undo the changes to these encode function and just have them stay fixed.

File: src/core/connection.h [hunk 17]
```diff
@@ -175,6 +175,16 @@ typedef union QUIC_CONNECTION_STATE {
...
 
+        //
+        // This flag indicates that the connection should ignore the fixed bit on received packets.
+        //
+        BOOLEAN IgnoreFixedBit : 1;
+
+        //
+        // This value of the fixed bit on send packets.
+        //
+        BOOLEAN FixedBit : 1;
```
Comment (nibanks) [inline src/core/connection.h:186]: See the comment at the top of this union. We need to keep this copy in sync with the one in quictypes.h.

File: src/core/binding.c (hunk 16 above)
Comment (liveans) [reply src/core/binding.c:1037]: Alright! I'm going to undo it to the original signature.

File: src/core/connection.c (hunk 15 above)
Comment (liveans) [reply src/core/connection.c:1969]: No, we don't need it. I'm gonna delete it in the next commit.

File: src/inc/msquic.h (hunk 14 above)
Comment (liveans) [reply src/inc/msquic.h:642]: I added stats, and I'll add them to the necessary places to increase values.

File: src/core/connection.c [hunk 18]
```diff
@@ -6655,6 +6661,9 @@ QuicConnGetV2Statistics(
     if (STATISTICS_HAS_FIELD(*StatsLength, DestCidUpdateCount)) {
         Stats->DestCidUpdateCount = Connection->Stats.Misc.DestCidUpdateCount;
     }
+    if(STATISTICS_HAS_FIELD(*StatsLength, RecvGreaseBitTpCount)) {
```
Comment (liveans) [inline src/core/connection.c:6664]:
```suggestion
    if (STATISTICS_HAS_FIELD(*StatsLength, RecvGreaseBitTpCount)) {
```

File: src/core/connection.h [hunk 19]
```diff
@@ -280,6 +290,7 @@ typedef struct QUIC_CONN_STATS {
 
         uint64_t TotalBytes;            // Sum of UDP payloads
         uint64_t TotalStreamBytes;      // Sum of stream payloads
+        uint64_t GreaseBitTpCount;      // Count of received grease transport parameter from peer.
```
Comment (liveans) [inline src/core/connection.h:293]: `uint64_t` might be overkill in here.

File: src/core/packet.h [hunk 20]
```diff
@@ -195,7 +195,7 @@ typedef struct QUIC_RETRY_PACKET_V1 {
 
     uint8_t UNUSED          : 4;
     uint8_t Type            : 2;
-    uint8_t FixedBit        : 1;    // Must be 1.
+    uint8_t FixedBit        : 1;    // Must be 1, unless grease_quic_bit tp has been sent.
```
Comment (liveans) [inline src/core/packet.h:198]:
```suggestion
    uint8_t FixedBit        : 1;    // Must be 1.
```

File: src/inc/msquic.h [hunk 21]
```diff
@@ -485,15 +485,17 @@ typedef struct QUIC_STATISTICS_V2 {
...
     // N.B. New fields must be appended to end
 
 } QUIC_STATISTICS_V2;
 
 #define QUIC_STRUCT_SIZE_THRU_FIELD(Struct, Field) \
     (FIELD_OFFSET(Struct, Field) + sizeof(((Struct*)0)->Field))
 
-#define QUIC_STATISTICS_V2_SIZE_1   QUIC_STRUCT_SIZE_THRU_FIELD(QUIC_STATISTICS_V2, KeyUpdateCount)         // v2.0 final size
-#define QUIC_STATISTICS_V2_SIZE_2   QUIC_STRUCT_SIZE_THRU_FIELD(QUIC_STATISTICS_V2, DestCidUpdateCount)     // v2.1 final size
+#define QUIC_STATISTICS_V2_SIZE_1   QUIC_STRUCT_SIZE_THRU_FIELD(QUIC_STATISTICS_V2, KeyUpdateCount)               // v2.0 final size
```
Comment (liveans) [inline src/inc/msquic.h:497]:
```suggestion
#define QUIC_STATISTICS_V2_SIZE_1   QUIC_STRUCT_SIZE_THRU_FIELD(QUIC_STATISTICS_V2, KeyUpdateCount)         // v2.0 final size
```

File: src/core/packet_builder.c [hunk 22]
```diff
@@ -197,6 +197,27 @@ QuicPacketBuilderPrepare(
...
+        (Connection->PeerTransportParams.Flags & QUIC_TP_FLAG_GREASE_QUIC_BIT) > 0) {
+        //
+        // Endpoints that receive the grease_quic_bit transport parameter from a peer SHOULD set the QUIC Bit to an unpredictable value
+        // unless another extension assigns specific meaning to the value of the bit.
+        //
+        uint8_t RandomValue;
+        (void) CxPlatRandom(sizeof(RandomValue), &RandomValue);
+        Connection->State.FixedBit = (RandomValue % 2);
+        Connection->State.FixedBitNegotiated = TRUE;
+        Connection->Stats.Recv.GreaseBitTpCount++;
```
Comment (liveans) [inline src/core/packet_builder.c:211]: I think this is not the right place to do it, any suggestion on it @nibanks?

File: src/inc/msquic.h [hunk 23]
```diff
@@ -485,6 +485,8 @@ typedef struct QUIC_STATISTICS_V2 {
 
     uint32_t DestCidUpdateCount;            // Number of times the destionation CID changed.
 
+    uint64_t RecvGreaseBitTpCount;          // Count of received grease transport parameter from the peer.
```
Comment (nibanks) [inline src/inc/msquic.h:488]: No, we don't need this. I was more saying we should have a flag at the top that says `GreaseBitNegotiated`.

Comment (liveans) [reply src/inc/msquic.h:488]: Oh, something like I added to stats? `FixedBitNegotiated`?

File: src/core/packet_builder.c [hunk 24]
```diff
@@ -197,6 +197,26 @@ QuicPacketBuilderPrepare(
...
+        (Connection->PeerTransportParams.Flags & QUIC_TP_FLAG_GREASE_QUIC_BIT) > 0) {
+        //
+        // Endpoints that receive the grease_quic_bit transport parameter from a peer SHOULD set the QUIC Bit to an unpredictable value
+        // unless another extension assigns specific meaning to the value of the bit.
+        //
+        uint8_t RandomValue;
+        (void) CxPlatRandom(sizeof(RandomValue), &RandomValue);
+        Connection->State.FixedBit = (RandomValue % 2);
+        Connection->Stats.GreaseBitNegotiated = TRUE;
+    }
```
Comment (nibanks) [inline src/core/packet_builder.c:211]: This logic should go in `QuicConnProcessPeerTransportParameters` which is called when we receive the transport parameters. Also please try to keep comments to 80 character width.

File: src/lib.rs [hunk 25]
```diff
@@ -459,14 +459,16 @@ pub struct QuicStatisticsRecv {
     /// Includes DuplicatePackets.
     pub dropped_packets: u64,
     pub duplicate_packets: u64,
-    /// Sum of UDP payloads
-    pub total_bytes: u64,
-    /// Sum of stream payloads
-    pub total_stream_bytes: u64,
```
Comment (nibanks) [inline src/lib.rs:465]: Why did you move these? From msquic.h:
```c
    struct {
        uint64_t TotalPackets;          // QUIC packets; could be coalesced into fewer UDP datagrams.
        uint64_t ReorderedPackets;      // Packets where packet number is less than highest seen.
        uint64_t DroppedPackets;        // Includes DuplicatePackets.
        uint64_t DuplicatePackets;
        uint64_t TotalBytes;            // Sum of UDP payloads
        uint64_t TotalStreamBytes;      // Sum of stream payloads
        uint64_t DecryptionFailures;    // Count of packet decryption failures.
        uint64_t ValidAckFrames;        // Count of receive ACK frames.
    } Recv;
```

File: src/inc/msquic.h [hunk 26]
```diff
@@ -448,6 +448,7 @@ typedef struct QUIC_STATISTICS_V2 {
     uint32_t StatelessRetry         : 1;
     uint32_t ResumptionAttempted    : 1;
     uint32_t ResumptionSucceeded    : 1;
+    uint32_t GreaseBitNegotiated    : 1;    // Set if we negotiated the GREASE bit.
```
Comment (nibanks) [inline src/inc/msquic.h:451]: Can you add a `uint32_t RESERVED : 27` too, so it's clear how much padding is here?

File: src/lib.rs (hunk 25 above)
Comment (liveans) [reply src/lib.rs:465]: Oh, looks like I got the structs mixed up

File: src/inc/msquic.h (hunk 26 above)
Comment (liveans) [reply src/inc/msquic.h:451]: Yes!

File: src/core/packet_builder.c (hunk 24 above)
Comment (liveans) [reply src/core/packet_builder.c]: I moved the logic into `QuicConnProcessPeerTransportParameters`, also I'm going to fix all comments with 80 character width!

File: src/core/connection.c [hunk 27]
```diff
@@ -2884,6 +2889,19 @@ QuicConnProcessPeerTransportParameters(
         "Peer Transport Parameters Set");
     Connection->State.PeerTransportParameterValid = TRUE;
 
+    if (Connection->Stats.GreaseBitNegotiated == FALSE &&
+        (Connection->PeerTransportParams.Flags & QUIC_TP_FLAG_GREASE_QUIC_BIT) > 0) {
```
Comment (nibanks) [inline src/core/connection.c:2893]: I don't think we should have this extra condition:
```suggestion
    if (Connection->PeerTransportParams.Flags & QUIC_TP_FLAG_GREASE_QUIC_BIT) {
```

Comment (nibanks) [reply src/core/connection.c:2893]: And actually, shouldn't we add a condition for only if the local setting is enabled?

Comment (nibanks) [reply src/core/connection.c:2893]: And, let's move this to the end of the function too please.

**Comment by nibanks** [comment]
Seems like we're getting this compiler warning/error too:
```
/home/vsts/work/1/s/src/core/crypto_tls.c:1192:1: error: function 'QuicCryptoTlsDecodeTransportParameters' exceeds recommended size/complexity thresholds [google-readability-function-size,-warnings-as-errors]
QuicCryptoTlsDecodeTransportParameters(
^
/home/vsts/work/1/s/src/core/crypto_tls.c:1192:1: note: 806 statements (threshold 800)
```
Can you go ahead and simply ignore that error?

File: src/core/connection.c (hunk 27 above)
Comment (liveans) [reply src/core/connection.c:2893]: > And actually, shouldn't we add a condition for only if the local setting is enabled?

Yes! Looks like I forgot it, I'll add it.

**Review by nibanks — DISMISSED** [review]

**Comment by csujedihy** [comment]
feel free to ignore [the arm docker build failure.](https://github.com/microsoft/msquic/runs/7925352774?check_suite_focus=true) for now.

**Review by nibanks — APPROVED** [review]

**Comment by nibanks** [comment]
Thanks so much for your contributions!!!
